components/identity-mgt-new/org.wso2.carbon.identity.mgt/src/main/java/org/wso2/carbon/identity/mgt/handler/internal/AccountLockEventHandler.java

```diff
@@ -0,0 +1,135 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.wso2.carbon.identity.mgt.handler.internal;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.base.IdentityException;
+import org.wso2.carbon.identity.core.model.IdentityErrorMsgContext;
+import org.wso2.carbon.identity.core.util.IdentityUtil;
+import org.wso2.carbon.identity.mgt.IdentityMgtConfig;
+import org.wso2.carbon.identity.mgt.constants.IdentityMgtConstants;
+import org.wso2.carbon.identity.mgt.dto.UserIdentityClaimsDO;
+import org.wso2.carbon.identity.mgt.event.IdentityMgtEvent;
+import org.wso2.carbon.identity.mgt.store.UserIdentityDataStore;
+import org.wso2.carbon.user.core.UserCoreConstants;
+import org.wso2.carbon.user.core.UserStoreException;
+import org.wso2.carbon.user.core.UserStoreManager;
+import org.wso2.carbon.user.core.util.UserCoreUtil;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+
+public class AccountLockEventHandler extends AbstractEventHandler {
+
+    private static final Log log = LogFactory.getLog(AccountLockEventHandler.class);
+
+    // the event types registered in this handler
+    protected ArrayList<String> registeredEventList;
+
+    @Override
+    public void init() {
+
+        registeredEventList = new ArrayList<String>() {{
+            add(IdentityMgtConstants.Event.PRE_AUTHENTICATION);
+        }};
+    }
+
+
+    @Override
+    public boolean isRegistered(IdentityMgtEvent event) {
+
+        HashMap<String, Object> properties = event.getEventProperties();
+        IdentityMgtConfig config = (IdentityMgtConfig) properties.get(IdentityMgtConstants.EventProperty.IDENTITY_MGT_CONFIG);
+
+        if (config.isAccountLockEnable()) {
+            //check whether the event name is exists in the registeredEventList
+            for (String eventName : registeredEventList) {
+                if (eventName.equals(event.getEventName())) {
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
+
+    @Override
+    public boolean handleEvent(IdentityMgtEvent identityMgtEvent) throws UserStoreException {
+
+        HashMap<String, Object> properties = identityMgtEvent.getEventProperties();
+
+        String userName = (String) properties.get(IdentityMgtConstants.EventProperty.USER_NAME);
+        UserStoreManager userStoreManager = (UserStoreManager) properties.get(IdentityMgtConstants.EventProperty.USER_STORE_MANAGER);
+        UserIdentityDataStore module = (UserIdentityDataStore) properties.get(IdentityMgtConstants.EventProperty.MODULE);
+        IdentityMgtConfig config = (IdentityMgtConfig) properties.get(IdentityMgtConstants.EventProperty.IDENTITY_MGT_CONFIG);
+
+        IdentityUtil.clearIdentityErrorMsg();
```

Comment:
This ThreadLocal variable was introduced because we were not able to propagate errors from the listener level to authenticators level, because the listener API return methods were either void or boolean. Detailed exception could not be sent. We need to think of a good way to send errors to layers above.

Shall we properly define error codes and descriptions for all the identity-mgt errors ?